docs/proposals/timeseries-partitioning-in-compactor.md

```diff
@@ -126,25 +126,27 @@ T1 - Partition 1-2 was created with hash % 2 == 0, and in order to avoid having
-6. Planner collects all compaction plans which are ready for compaction.
-   1. For each compaction plan and for each blocks in the plan:
+      3. Generates compaction group that indicates which partition ID each blocks got assigned.
+      4. Partitioned compaction group information would be stored in block storage in order for cleaner to pick it up later.
+4. Grouper returns compaction groups to Compactor.
+5. Compactor iterates over each partitioned compaction group. For each iteration, calls Planner to make sure the group is ready for compaction.
+6. Planner collects all compaction groups which are ready for compaction.
+   1. For each compaction group and for each blocks in the plan:
       1. Make sure the source block fits within the time range of the group.
-      2. Make sure the source block with assigned partition ID is currently not used by another ongoing compaction.
+      2. Make sure the source block with assigned partition ID is currently not used by another ongoing compaction. This could utilize visit marker file that is introduced in #4805 by expanding it for each partition ID of the source block.
    2. If all blocks in the group are ready to be compacted, mark status of those blocks along with assigned partition ID as `pending`.
-7. Return all ready compaction plans to Compactor.
-8. Compactor starts compacting partitioned blocks from compaction plans. Once compaction completed, Compactor would upload deletion marker for the source blocks in the plan. Also, it would mark status of all blocks along with assigned partition ID in the group as `completed`.
+   3. The status information of each partition ID would be stored in block storage in order for cleaner to pick it up later.
+7. Return all ready compaction groups to Compactor.
+8. Compactor starts compacting partitioned blocks from compaction groups. Once compaction completed, Compactor would upload deletion marker for the source blocks in the plan. Also, it would mark status of all blocks along with assigned partition ID in the group as `completed`.
 
 ### Clean up Workflow
 
-Cleaner would periodically check any source blocks having a deletion marker. If there is a deletion marker for the block, Cleaner should retrieve compaction plan to get all partition IDs assigned to this block. If this source block's all assigned partition ID have status set to `completed`, this source block can be deleted. Otherwise, skip deletion.
+Cleaner would periodically check any source blocks having a deletion marker. If there is a deletion marker for the block, Cleaner should retrieve compaction group information from block storage to get all partition IDs assigned to this block as well as status information of each partition ID. If this source block's all assigned partition ID have status set to `completed`, this source block can be deleted. Otherwise, skip deletion.
```

Comment:
If we store information at the block storage, do we have any concern about consistency? S3 does support strong read after write consistency but I think other object storage we support might not have the same feature.

What about relying on some KV stores for such kind of metadata?

Reply:
I asked this question because I think it is still not very clear to me how we want to store the information. We use a plan file per compactor, or different files for every compaction group per compactor? 

Considering the consistency and file immutability, I feel this is something we need to think about. We don't need to dive too deep, but a few more details would be good. Like have we considering using KV stores for such metadata usecase?

Reply:
I updated the doc. `partitioned compaction group` file would be stored under tenant level in block storage. Each `partitioned compaction group` has an ID that is hash of min and max time of result block. Each `partitioned compaction group` contains a list of partitions. Each parition contains partition ID and list of blocks used in this partition. Check Compaction Partitioning Examples, each scenario generates one `partitioned compaction group` and its partitions.

The original proposal is based on shuffle sharding compactor (grouper and planner) and it is mainly based on S3 as block storage. In order to not introduce complexity, I choose block storage to store those information.